File: Application/src/main/java/com/teeh/klimasensor/common/ts/SimpleTs.java
package com.teeh.klimasensor.common.ts;

import com.teeh.klimasensor.TsEntry;
import com.teeh.klimasensor.common.utils.CalcUtil;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Created by teeh on 16.07.2017.
 */

public class SimpleTs {

    public static final DateFormat tsFormat = new SimpleDateFormat("yyyy-MM-dd kk:mm:ss", Locale.ENGLISH);

    private List<SimpleEntry> data;
    private Map<Date, Integer> dateMap;

    public SimpleTs(List<SimpleEntry> data_) {
        data = data_;
        dateMap = new HashMap<Date, Integer>();

        for (int i=0; i<data.size(); i++) {
            dateMap.put(data.get(i).getTimestamp(), i);
        }
    }

    public List<SimpleEntry> getTs(){
        return data;
    }

    public SimpleEntry getEntry(Date d) {
        Integer index = dateMap.get(d);
        if (index != null && index < data.size()) {
            return data.get(index);
        }
        else {
            return null;
        }
    }

    public List<Double> getValueList() {
        return data.stream().map(x -> x.getValue()).collect(Collectors.toList());
    }


    ////////////
    // Limits //
    ////////////

    public Double getMin() {
        return Collections.min(getValueList());
    }

    public Double getMax() {
        return Collections.max(getValueList());
    }

    public SimpleEntry getLatestEntry() {
        return data.get(data.size()-1);
    }


    /////////////////
    // Aggregators //
    /////////////////

    public Double getLatestValue() {
        return getLatestEntry().getValue();
    }

    public Double getAvg() {
        List<Double> ts = getValueList();
        Double sum = CalcUtil.sum(ts);
        return sum / ts.size();
    }

    public Double getMedian() {
        List<Double> ts = getValueList();
        Collections.sort(ts);
        if (ts.size() % 2 == 0) {
            int half = ts.size()/2;
            return (ts.get(half-1) + ts.get(half)) / 2.0;
        }
        else {
            int half = (ts.size()-1)/2;
            return ts.get(half);
        }
    }

    public Date getLatestTimestamp() {
        if (data == null || data.isEmpty()) {
            return null;
        }
        else {
            SimpleEntry lastEntry = data.get(data.size()-1);
            return lastEntry.getTimestamp();
        }
    }

    public Date getFirstTimestamp() {
        if (data == null || data.isEmpty()) {
            return null;
        }
        else {
            SimpleEntry firstEntry = data.get(0);
            return firstEntry.getTimestamp();
        }
    }

    public String getLatestTimestampString() {
        Date lastTs = getLatestTimestamp();
        if (lastTs == null) {
            return "1970-01-01 00:00:00";
        } else {
            return tsFormat.format(lastTs);
        }
    }
}

File: Application/src/main/java/com/teeh/klimasensor/common/ts/SimpleTs.kt
package com.teeh.klimasensor.common.ts

import com.teeh.klimasensor.common.utils.CalcUtil
import com.teeh.klimasensor.common.utils.DateUtils
import java.time.LocalDate
import java.time.LocalDateTime

import java.util.Date
import java.util.HashMap

class SimpleTs(val ts: List<SimpleEntry>) {
    private val dateMap: MutableMap<LocalDateTime, Int>

    val valueList: List<Double?>
        get() = ts.map { x -> x.value }

    ////////////
    // Limits //
    ////////////

    val min: Double?
        get() = valueList.filterNotNull().min()

    val max: Double?
        get() = valueList.filterNotNull().max()

    val latestEntry: SimpleEntry
        get() = ts.last()


    /////////////////
    // Aggregators //
    /////////////////

    val latestValue: Double?
        get() = latestEntry.value

    val avg: Double
        get() {
            val ts = valueList
            val sum = CalcUtil.sum(ts)
            return sum / ts.size
        }

    val median: Double
        get() {
            val ts = valueList.filterNotNull().sorted()
            if (ts.size % 2 == 0) {
                val half = ts.size / 2
                return (ts[half - 1] + ts[half]) / 2.0
            } else {
                val half = (ts.size - 1) / 2
                return ts[half]
            }
        }

    val latestTimestamp: LocalDateTime
        get() = latestEntry.timestamp

    val firstTimestamp: LocalDateTime
        get() = ts[0].timestamp

    val latestTimestampString: String
        get() = DateUtils.toString(latestTimestamp)

    val firstTimestampString: String
        get() = DateUtils.toString(firstTimestamp)


    init {
        dateMap = HashMap()
        ts.forEachIndexed { index, simpleEntry -> dateMap.put(simpleEntry.timestamp, index) }
    }

    fun getEntry(d: LocalDateTime): SimpleEntry? {
        val index = dateMap[d]
        return if (index != null && index < ts.size) {
            ts[index]
        } else {
            null
        }
    }
}
